covid19_sfbayarea/data/solano.py

```diff
@@ -0,0 +1,286 @@
+                    'outFields': 'date_reported,cumulative_number_of_cases_on_t,total_deaths,residents_tested,new_cases_confirmed_today',
+                    'orderByFields': 'date_reported asc', 'f': 'json'}
+    response = requests.get(data_url, params=param_list)
+    response.raise_for_status()
+    parsed = response.json()
+    features = [obj["attributes"] for obj in parsed['features']]
+
+    # convert dates
+    PACIFIC_TIME = dateutil.tz.gettz('America/Los_Angeles')
+    for obj in features:
+        timestamp = obj['date_reported']
+        date = datetime.fromtimestamp(timestamp/1000, tz=PACIFIC_TIME)
+        obj['date_reported'] = date.strftime("%Y-%m-%d")
+
+    re_keyed = [{TIMESERIES_KEYS[key]: value for key, value in entry.items()}
+                for entry in features]
+
+    # Templates have all data points in the data model
+    # datapoints that are not reported have a value of -1
+    CASES_TEMPLATE = { "date": -1, "cases":-1, "cumul_cases":-1 }
+    DEATHS_TEMPLATE = {"date":-1, "deaths":-1, "cumul_deaths":-1 }
+    TESTS_TEMPLATE = { "date": -1, "tests": -1, "positive": -1, "negative": -1, "pending": -1, "cumul_tests": -1,"cumul_pos": -1, "cumul_neg": -1, "cumul_pend": -1 }
+
+    for entry in re_keyed:
+        #FIXME:
+        # Sonoma county has entries for each day, but some days have "null" for cumulative deaths and cumulative tests.
+        # This means that cumulative counts go something like: 1, null ,null, null, 2, 3 -- on consecutive days.
+        # My best guess is that for deaths and tests they are entering "null" for the cumulative counts on days where the cumulative counts haven't changed from the previous day.
+        # The solution below is just to filter out the cumulative count "null" values (Python None), which results in gaps in the timeseries.
+        # We may want to revisit this and clean up our json to interpolate the values on null-count days until the day the value changes. But that means we're injecting numbers
+        # into the data, and I don't know if we want to do that. But we will be interpolating anyways, in our visualizations, so maybe that's ok.
+
+        # grab cumulative values, to check if they are None
+        cumul_cases = entry["cumul_cases"]
+        cumul_deaths = entry["cumul_deaths"]
+        cumul_tests = entry["cumul_tests"]
+
+        # grab cases and replace None with 0
+        if entry["cases"] is None:
+            entry["cases"] = 0
+
+        if cumul_cases is not None:
+            cases_entry = { k: entry[k] for k in CASES_TEMPLATE if k in entry }
+            series["cases"].append(cases_entry)
+        if cumul_deaths is not None:
+            deaths_entry = {k: entry[k] for k in DEATHS_TEMPLATE if k in entry}
+            series["deaths"].append(deaths_entry)
+        if cumul_tests is not None:
+            tests_entry = {k: entry[k] for k in TESTS_TEMPLATE if k in entry}
+            series["tests"].append(tests_entry)
```

Comment:
I agree with you, in principle. In the case of the `tests` series for Solano, doing the inverse and extrapolating daily counts from reported cumulative counts means that the first day tests were reported, it looks like we'd be saying there were 1855 tests administered or results confirmed on 4/13, which I doubt. I think the visualizations are relying on cumulative counts, so that's why I'm ok with the asymmetry of extrapolating cumulative  counts from daily counts  but not the inverse. I could be convinced otherwise though, I don't think it's a big deal if we note it in the README.

Reply:
Argh, I also just noticed that the source data calls this 'Residents Tested' and the dashboard  calls this "Number of Residents Tested" which to my mind means they might not be counting individuals getting retested. Whereas other counties are simply reporting the tests. That makes me even more reluctant to do anything more than report exactly what they report.

Reply:
@ldtcooper @kwonangela7 Thoughts?

Reply:
> however, it looks like we'd be saying there were 1855 on 4/13, which I doubt.

Oh, good point. What if we filled in the daily counts on all days after the first, and left the first at `-1` or `null`? Dunno if better or not, but at least somewhat more convenient for consumers.

> I also just noticed that the source data calls this 'Residents Tested' and the dashboard calls this "Number of Residents Tested" which to my mind means they might not be counting individuals getting retested.

😫